src/storage/src/source/reclock.rs

```diff
@@ -0,0 +1,422 @@
+
+//! Timestamper using persistent collection
+use std::collections::{HashMap, VecDeque};
+use std::pin::Pin;
+use std::time::Duration;
+
+use anyhow::{bail, Context as _};
+use differential_dataflow::lattice::Lattice;
+use futures::{Stream, StreamExt};
+use mz_dataflow_types::client::controller::storage::CollectionMetadata;
+use timely::progress::{Antichain, Timestamp as _};
+use timely::PartialOrder;
+
+use mz_dataflow_types::sources::MzOffset;
+use mz_expr::PartitionId;
+use mz_ore::now::NowFn;
+use mz_persist_client::read::ListenEvent;
+use mz_persist_client::write::WriteHandle;
+use mz_persist_client::Upper;
+use mz_repr::{Diff, Timestamp};
+use tracing::{error, info};
+
+pub struct ReclockOperator {
+    name: String,
+    // How far on the persist collection we've read
+    read_progress: Antichain<Timestamp>,
+    // Upper for the underlying persist collection
+    write_upper: Antichain<Timestamp>,
+    // If a binding is here, it has been persisted.  Previously returned bindings are compacted.
+    persisted_timestamp_bindings: HashMap<PartitionId, VecDeque<(Timestamp, MzOffset)>>,
+    // Max offset we've returned (or have committed to returning at the end of the current invocation).
+    read_cursors: HashMap<PartitionId, MzOffset>,
+    write_handle: WriteHandle<(), PartitionId, Timestamp, Diff>,
+    timestamp_bindings_listener:
+        Pin<Box<dyn Stream<Item = ListenEvent<(), PartitionId, Timestamp, Diff>>>>,
+    now: NowFn,
+    update_interval: Duration,
+    last_timestamp: Timestamp,
+}
+
+impl ReclockOperator {
+    pub async fn new(
+        name: String,
+        CollectionMetadata {
+            persist_location,
+            timestamp_shard_id,
+        }: CollectionMetadata,
+        now: NowFn,
+        update_interval: Duration,
+        mut as_of: Antichain<Timestamp>,
+    ) -> anyhow::Result<Self> {
+        let persist_client = persist_location
+            .open()
+            .await
+            .with_context(|| "error creating persist client")?;
+
+        let (write_handle, read_handle) = persist_client
+            .open(timestamp_shard_id)
+            .await
+            .expect("persist handles open err");
+
+        let read_progress = read_handle.since().clone();
+        let write_upper = write_handle.upper().clone();
+
+        // This is an invariant of the persist collection.
+        assert!(
+            timely::PartialOrder::less_equal(&read_progress, &write_upper),
+            "{:?} PARTIAL ORDER: {:?} {:?}",
+            name,
+            read_progress,
+            write_upper,
+        );
+
+        if timely::PartialOrder::less_than(&as_of, &read_progress) {
+            bail!(
+                "Source AS OF must not be less than since: {:?} vs {:?}",
+                as_of,
+                read_progress
+            );
+        }
+
+        // XXX: what do we do if the write_upper is less than the as of? e.g. if we get a non-zero as of on a new source?
```

Comment:
flagging this! I'm not sure what the right answer is.

Reply:
I think the correct behaviour is to assert this is never the case because for this to happen the storage controller would have to pick an `as_of` value that is in the future of the write upper that it knows about. We can't really recover from this error and it's also an indication that something is very very wrong

Reply:
okay cool.  I'll have it panic if as_of < write_upper AND write_upper != 0